components/script/serviceworkerjob.rs

```diff
@@ -119,52 +130,68 @@ impl AsyncJobHandler {
 impl Runnable for AsyncJobHandler {
     #[allow(unrooted_must_root)]
     fn main_thread_handler(self: Box<AsyncJobHandler>, script_thread: &ScriptThread) {
-        // NOTE: need to clone scope as self.promise.root() consumes self, below, so
-        // we can't access `self.scope_url`
-        let scope_url = self.scope_url.clone();
-        let settle_type = self.settle_type.clone();
-        let global = self.global.root();
-        let promise = self.promise.root();
-        let _ac = JSAutoCompartment::new((&*global).get_cx(), promise.reflector().get_jsobject().get());
-        match settle_type {
-            SettleType::Resolve(reg) => {
-                let reg = reg.root();
-                promise.resolve_native((&*global).get_cx(), &*reg);
-                script_thread.finish_job(&scope_url, &*global);
-            }
-            SettleType::Reject(err_type) => {
-                promise.reject_error((&*global).get_cx(), err_type);
-                script_thread.finish_job(&scope_url, &*global);
+        use self::InvokeType::*;
+        match self.invoke_type {
+            Run => script_thread.invoke_run_job(self),
+            Register => script_thread.invoke_register_job(self),
+            _ => {
+                let handler = *self;
+                let scope_url = handler.scope_url;
+                let invoke_type = handler.invoke_type;
+                let trusted_global = handler.global;
+                let global = trusted_global.root();
+                let promise = &*handler.promise.root();
+                let _ac = JSAutoCompartment::new((&*global).get_cx(), promise.reflector().get_jsobject().get());
+                match invoke_type {
+                    Settle(SettleType::Resolve(reg)) => {
+                        let reg = reg.root();
+                        promise.resolve_native((&*global).get_cx(), &*reg);
+                        let finish_job_handler = box FinishJobHandler {
+                            scope_url: scope_url,
+                            global: trusted_global
+                        };
+                        script_thread.queue_finish_job(finish_job_handler, &*global);
+                    }
+                    Settle(SettleType::Reject(err_type)) => {
+                        promise.reject_error((&*global).get_cx(), err_type);
+                        let finish_job_handler = box FinishJobHandler {
+                            scope_url: scope_url,
+                            global: trusted_global
+                        };
+                        script_thread.queue_finish_job(finish_job_handler, &*global);
```

Comment:
This code is duplicated between the arms, so we can move it into code that runs afterwards instead.